plugins/outputs/sematext/serializer/serializer.go

```diff
@@ -156,7 +166,98 @@ func serializeMetric(key string, value interface{}) string {
 // repeating the same tags+timestamp multiple times.
-// TODO to be implemented in the future to make Telegraf requests to Sematext backend smaller
 type CompactMetricSerializer struct {
-	tagsIDToMetrics map[string]telegraf.Metric
+	log    telegraf.Logger
+	config *serializationConfig
+}
+
+// NewCompactMetricSerializer creates an instance of CompactMetricSerializer
+func NewCompactMetricSerializer(log telegraf.Logger) *CompactMetricSerializer {
+	return &CompactMetricSerializer{
+		log:    log,
+		config: NewSerializationConfig(),
+	}
+}
+
+// Write serializes input metrics array according to Sematext variant of influx line protocol. The output is returned
+// as []byte which can be empty if there were no metrics or metrics couldn't be serialized.
+func (s *CompactMetricSerializer) Write(metrics []telegraf.Metric) []byte {
+	var output bytes.Buffer
+	idToMetrics := make(map[string][]telegraf.Metric)
+
+	// first group the metrics that share the same identification
+	for _, m := range metrics {
+		id := buildId(m)
+		idToMetrics[id] = append(idToMetrics[id], m)
+	}
+
+	// sort the keys keep the order fixed
+	sortedIds := make([]string, 0, len(idToMetrics))
+	for i := range idToMetrics {
+		sortedIds = append(sortedIds, i)
+	}
+	sort.Strings(sortedIds)
+
+	// then create 1 metrics line for each of created groups
+	for _, groupId := range sortedIds {
+		metrics := idToMetrics[groupId]
+		serializedTags := serializeTags(metrics[0].Tags(), s.config.notSerializable)
+		serializedMetrics := serializeMetrics(metrics)
+		serializedTimestamp := strconv.FormatInt(metrics[0].Time().UnixNano(), 10)
+
+		if serializedMetrics == "" {
+			continue
+		}
+
+		output.WriteString(nameEscape(metrics[0].Name()))
+		if serializedTags != "" {
+			output.WriteString(",")
+			output.WriteString(serializedTags)
+		}
+		output.WriteString(" ")
+		output.WriteString(serializedMetrics)
+		output.WriteString(" ")
+		output.WriteString(serializedTimestamp)
+		// has to end with a newline
+		output.WriteString("\n")
+	}
+
+	return output.Bytes()
+}
+
+func serializeMetrics(metrics []telegraf.Metric) string {
+	var serializedMetrics strings.Builder
+	fieldList := make([]*telegraf.Field, 0)
+
+	for _, metric := range metrics {
+		for _, field := range metric.FieldList() {
+			fieldList = append(fieldList, field)
+		}
+	}
+
+	// make the field order sorted
+	sort.Slice(fieldList, func(i, j int) bool {
+		return fieldList[i].Key < fieldList[j].Key
+	})
+
+	var countAdded = 0
+	for _, field := range fieldList {
+		var serializedMetric = serializeMetricField(field.Key, field.Value)
+
+		if serializedMetric == "" {
+			continue
+		}
+
+		if countAdded > 0 {
+			serializedMetrics.WriteString(",")
+		}
+		serializedMetrics.WriteString(serializedMetric)
+		countAdded++
+	}
+
+	return serializedMetrics.String()
+}
+
+func buildId(metric telegraf.Metric) string {
+	return fmt.Sprint(metric.Time().UnixNano()) + "-" + metric.Name() + "-" + tags.GetTagsKey(metric.Tags())
```

Comment:
Maybe use `strconv.FormatInt()`  to convert the timestamp to string and avoid using fmt.Sprint to reduce allocations?

Reply:
Aha.